demos/dev_mode_key_provisioning/src/aws_dev_mode_key_provisioning.c

```diff
@@ -788,28 +987,52 @@ CK_RV xProvisionDevice( CK_SESSION_HANDLE xSession,
         }
     }
 
-    /* Ensure that the above procedure has ended up with a client certificate and
-     * private key available in storage. */
+    /* Check whether a keyset is now present. In order to support X.509
+     * certificate enrollment, the public and private key objects must both be
+     * available. */
     if( xResult == CKR_OK )
     {
-        for( ; ulObjectIndex < sizeof( pxPkcsLabels ) / sizeof( pxPkcsLabels[ 0 ] ); ulObjectIndex++ )
+        xResult = xGetProvisionedState( xSession,
+                                        pxFunctionList,
+                                        &xProvisionedState );
+
+        if( ( 0 == xProvisionedState.xPrivateKey ) || ( 0 == xProvisionedState.xPublicKey ) )
         {
-            xResult = xFindObjectWithLabelAndClass( xSession,
-                                                    ( const char * ) pxPkcsLabels[ ulObjectIndex ],
-                                                    xClass[ ulObjectIndex ],
-                                                    &xObject );
+            /* Generate a new keyset if either of the above objects couldn't be
+             * found. */
+            xResult = xProvisionGenerateKeyPairEC( xSession,
+                                                   ( uint8_t * ) pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS,
+                                                   ( uint8_t * ) pkcs11configLABEL_DEVICE_PUBLIC_KEY_FOR_TLS,
+                                                   &xProvisionedState.xPrivateKey,
+                                                   &xProvisionedState.xPublicKey );
+        }
 
-            if( xResult != CKR_OK )
-            {
-                configPRINTF( ( "Error: required crypto object %s is missing.\r\n", pxPkcsLabels[ ulObjectIndex ] ) );
-                break;
-            }
+        /* Ensure that an error condition is set if either object is still
+         * missing. */
+        if( ( CKR_OK == xResult ) &&
+            ( ( 0 == xProvisionedState.xPrivateKey ) ||
+              ( 0 == xProvisionedState.xPublicKey ) ) )
+        {
+            xResult = CKR_KEY_HANDLE_INVALID;
         }
     }
 
+    /* Log the device public key for developer enrollment purposes, but only if
+     * there's not already a certificate. */
+    if( ( CKR_OK == xResult ) && ( 0 == xProvisionedState.xClientCertificate ) )
+    {
+        xResult = xWritePublicKeyToConsole( xSession,
+                                            pxFunctionList,
+                                            xProvisionedState.xPublicKey );
```

Comment:
Even though it is already covered in the README, since users who forget to paste their client credentials (or pop their stash!) will end up here unintentionally, I think it would be good to print a reminder to the console reminding people that the demo won't work just yet and that they still need to create a certificate.